src/adapter.rs

```diff
@@ -0,0 +1,739 @@
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+    fn get_pairable_timeout(&self) -> Result<u32, Box<Error>>;
+    fn set_pairable_timeout(&self, timeout: u32) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+    fn get_discoverable_timeout(&self)-> Result<u32, Box<Error>>;
+    fn set_discoverable_timeout(&self, timeout: u32) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    } 
+    fn is_discovering(&self)-> Result<bool, Box<Error>>;
+    fn set_discovering(&self, discovering: bool) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    } 
+    fn set_can_start_discovery(&self, can_start_discovery: bool) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+    fn set_can_stop_discovery(&self, can_stop_discovery: bool) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+    fn create_discovery_session(&self)-> Result<BluetoothDiscoverySession, Box<Error>> ;
+    fn get_uuids(&self)-> Result<Vec<String>, Box<Error>>;
+    fn set_uuids(&self, uuids: Vec<String>) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    } 
+    fn get_vendor_id_source(&self)-> Result<String, Box<Error>>;
+    fn get_vendor_id(&self)-> Result<u32, Box<Error>>;
+    fn get_product_id(&self) -> Result<u32, Box<Error>> ;
+    fn get_device_id(&self) -> Result<u32, Box<Error>>;
+    fn get_modalias(&self) -> Result<(String, u32, u32, u32), Box<Error>>;  
+    fn set_modalias(&self, modalias: String) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+    fn get_ad_datas(&self) -> Result<Vec<String>, Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+    fn set_ad_datas(&self, ad_datas: Vec<String>) -> Result<(), Box<Error>>{
+        Err(Box::from(NOT_SUPPORTED_ERROR))  
+    }
+}
+
+impl BluetoothAdapter{
+    #[cfg(all(target_os = "linux", feature = "bluetooth"))]
+    pub fn new() -> Result<Box<BluetoothAdapterBluez>, Box<Error>> {
+        let bluez_adapter = try!(BluetoothAdapterBluez::init());
+        Ok(Box::new(bluez_adapter))
+    }
+
+    #[cfg(all(target_os = "android", feature = "bluetooth"))]
+    pub fn new() -> Result<Box<BluetoothAdapterAndroid>, Box<Error>> {
+        let blurdroid_adapter = try!(BluetoothAdapterAndroid::get_adapter());
+        Ok(Box::new(blurdroid_adapter))
+    }
+
+    #[cfg(all(target_os = "macos", feature = "bluetooth"))]
+    pub fn new() -> Result<Box<BluetoothAdapterMac>, Box<Error>> {
+        let mac_adapter = try!(BluetoothAdapterMac::init());
+        Ok(Box::new(mac_adapter))
+    }
+
+    #[cfg(not(any(all(target_os = "linux", feature = "bluetooth"),
+                  all(target_os = "android", feature = "bluetooth"),
+                  all(target_os = "macos", feature = "bluetooth"))))]
+    pub fn new() -> Result<Box<BluetoothAdapterEmpty>, Box<Error>> {
```

Comment:
All of these new methods should be returning `Box<BluetoothAdapter>`, not a specific per-platform type.

Reply:
We initially tried returning Box<BluetoothAdapter>, but the we found that empty::BluetoothAdapter was not expecting a trait. 
We kept getting the following error,
adapter::BluetoothAdapter is not implemented for empty::BluetoothAdapter

Reply:
Yes. That means that you need to implement the BluetoothAdapter trait for the empty bluetooth adapter type in empty.rs.

Reply:
Am I right in assuming that the struct BluetoothAdapter in empty.rs needs to be changed to a trait as we did in adapter.rs?

Reply:
With respect to this I have a few doubts?

1) If  try making empty : : bluetoothadapter into a trait in empty.rs, I can no longer have 
struct Empty(Arc<BluetoothAdapterEmpty) in adapter.rs. Because
BluetoothAdapterEmpty cannot be made into an object if it is a trait and it will give rise to more errors with the implementation now present in adapter.rs?

2) init() and new() present in empty::BluetoothAdapter empty.rs cannot be part of the new trait implementation?

Reply:
It needs to be part of the `impl BluetoothAdapter`, rather than `trait BluetoothAdapter`.

Reply:
Actually, it's quite possible that I am misunderstanding what you have tried to do. The struct in empty.rs should remain a struct, because it is a concrete type that is expected to implement the BluetoothAdapter trait that you already created.

Reply:
It's no different than the various other adapter structs that implement the trait.

Reply:
> Yes. That means that you need to implement the BluetoothAdapter trait for the empty bluetooth adapter type in empty.rs.

So if I have to leave the struct in empty.rs as such, what should I change in empty.rs?

> It needs to be part of the `impl BluetoothAdapter`, rather than `trait BluetoothAdapter`.
Yes, the method new() is in impl BluetoothAdapter as of our current implementation. 
But when the return type is Box'BluetoothAdapter' for fn new() in adapter.rs , it says the BluetoothAdapter in empty.rs does not expect a trait.
This got me a little confused

Reply:
I recommend renaming the struct in empty.rs to EmptyAdapter for clarity. Just like there is `impl BluetoothAdapter for Mac` (for example), we will also need to implement the BluetoothAdapter trait for the EmptyAdapter struct. When this implementation exists, it will be possible to return an instance of `Box<EmptyAdapter>` in a function the returns the `Box<BluetoothAdapter>` type. Does that make sense?

Reply:
Yes. Let me make sure I am got it right.
In **empty.rs**,
Define a trait BlueoothAdapter, rename struct to EmptyAdapter, define all functions under impl BluetoothAdapter for EmptyAdapter, but define init() and new() functions separately in impl BluetoothAdapter.

Reply:
There's no need to define a new trait. There is already a BluetoothAdapter trait that you have defined in this PR.

Reply:
Okay, Thank you!
But the init() function in empty.rs should be placed in impl BluetoothAdapter for EmptyAdapter right? As fn new() in adapter.rs expects a trait. Which also means I have to add it as part trait BluetoothAdapter?

Reply:
I do not believe that change is required. The initial method cam conitinue to leave be in an `impl EmptyAdapter` block.

Reply:
We fixed it for empty